File: Sources/InstantTranslator/Models/LicenseState.swift
import Foundation

enum LicenseState {
    case checking
    case trial(daysRemaining: Int)
    case expired
    case active(planName: String?, validatedAt: Date, isOfflineGrace: Bool)
    case invalid(reason: String)

    var canTranslate: Bool {
        switch self {
        case .trial, .active:
            true
        case .checking, .expired, .invalid:
            false
        }
    }

    var statusText: String {
        switch self {
        case .checking:
            return "Checking license"
        case .trial(let days):
            return "Trial: \(days) days remaining"
        case .expired:
            return "Trial expired"
        case .active(let planName, _, let isOfflineGrace):
            let base: String
            if let planName, !planName.isEmpty {
                base = "Active: \(planName)"
            } else {
                base = "Active license"
            }
            return isOfflineGrace ? "\(base) (offline mode)" : base
        case .invalid(let reason):
            return "Invalid license: \(reason)"
        }
    }
}

File: windows/ControlV.Core/Models/LicenseState.cs
namespace ControlV.Core.Models;

/// Mirrors Swift `LicenseState` (an enum with payloads) as a closed record hierarchy.
public abstract record LicenseState
{
    public sealed record Checking : LicenseState;
    public sealed record Trial(int DaysRemaining) : LicenseState;
    public sealed record Expired : LicenseState;
    public sealed record Active(string? PlanName, DateTimeOffset ValidatedAt, bool IsOfflineGrace) : LicenseState;
    public sealed record Invalid(string Reason) : LicenseState;

    public bool CanTranslate => this is Trial or Active;

    public string StatusText => this switch
    {
        Checking => "Checking license",
        Trial t => $"Trial: {t.DaysRemaining} days remaining",
        Expired => "Trial expired",
        Active a => (string.IsNullOrEmpty(a.PlanName) ? "Active license" : $"Active: {a.PlanName}") + (a.IsOfflineGrace ? " (offline mode)" : ""),
        Invalid i => $"Invalid license: {i.Reason}",
        _ => "Unknown",
    };
}
